sdk/go/property/path.go

```diff
@@ -0,0 +1,215 @@
+			return Value{}, PathApplyIndexOutOfBounds{found: v.AsArray(), idx: i.int}
+		}
+		// We now want to set set v.AsArray()[i]=to, but we don't want to mutate v. We make
+		// a shallow copy of v and then alter that.
+		cp := make(Array, len(v.AsArray()))
+		copy(cp, v.AsArray())
+		cp[i.int] = to
+		return butLast.Set(src, New(cp)) // Finally, propagate the shallow change back up.
+	case v.IsMap():
+		k, ok := last.(KeySegment)
+		if !ok {
+			return Value{}, pathErrorf(v, "expected a key step, found %T", last)
+		}
+		// We now want to set set v.AsMap()[k]=to, but we don't want to mutate v. We make
+		// a shallow copy of v and then alter that.
+		cp := make(Map, len(v.AsMap()))
+		for k, e := range v.AsMap() {
+			cp[k] = e
+		}
+		cp[k.string] = to                // Assign to the new map
+		return butLast.Set(src, New(cp)) // Finally, propagate the shallow change back up.
+	default:
+		return Value{}, pathErrorf(v, "expected a map or array, found %s", typeString(v))
+	}
+}
+
+// Alter changes the value at p by applying f.
+//
+// To preserve metadata, use [WithGoValue] in conjunction with Alter:
+//
+//	p.Alter(v, func(v Value) Value) {
+//		return property.WithGoValue(v, "new-value")
+//	})
+//
+// This will preserve any secrets or dependencies encoded in `v`.
+func (p Path) Alter(v Value, f func(v Value) Value) (Value, PathApplyFailure) {
+	oldValue, err := p.Get(v)
+	if err != nil {
+		return Value{}, err
+	}
+	return p.Set(v, f(oldValue))
+}
+
+type PathSegment interface {
+	apply(Value) (Value, PathApplyFailure)
+}
+
+// NewSegment creates a new [PathSegment] suitable for use in [Path].
+func NewSegment[T interface{ string | int }](v T) PathSegment {
+	switch v := any(v).(type) {
+	case string:
+		return KeySegment{v}
+	case int:
+		return IndexSegment{v}
+	default:
+		panic("impossible")
+	}
+}
+
+type PathApplyFailure interface {
+	error
+
+	Found() Value
+}
+
+// KeySegment represents a traversal into a map by a key.
+//
+// KeySegment does not support glob ("*") expansion. Values are treated as is.
+type KeySegment struct{ string }
+
+func (k KeySegment) apply(v Value) (Value, PathApplyFailure) {
+	if v.IsMap() {
+		m := v.AsMap()
+		r, ok := m[k.string]
+		if ok {
+			return r, nil
+		}
+		return Value{}, PathApplyKeyMissing{found: m, needle: k.string}
+	}
+	return Value{}, PathApplyKeyExpectedMap{found: v}
+}
+
+type IndexSegment struct{ int }
+
+func (k IndexSegment) apply(v Value) (Value, PathApplyFailure) {
+	if v.IsArray() {
+		a := v.AsArray()
+		if k.int < 0 || k.int >= len(a) {
+			return Value{}, PathApplyIndexOutOfBounds{found: a, idx: k.int}
+		}
+		return a[k.int], nil
+	}
+	return Value{}, PathApplyIndexExpectedArray{found: v}
+}
+
+type PathApplyKeyExpectedMap struct {
+	found Value
+}
+
+func (err PathApplyKeyExpectedMap) Error() string {
+	return "expected a map, found a " + typeString(err.found)
+}
+
+func (err PathApplyKeyExpectedMap) Found() Value {
+	return err.found
+}
+
+type PathApplyKeyMissing struct {
```

Comment:
I think generally we suffix Errors with `Error`.  Should we do that for all the errors here as well?

Also how are we expecting these error types to be used?